subsys/net/lib/nrf_cloud/src/nrf_cloud_logs.c

```diff
@@ -0,0 +1,452 @@
+	} /* Else send ring buffer contents as is -- JSON or raw binary. */
+
+	if (IS_ENABLED(CONFIG_NRF_CLOUD_MQTT)) {
+		err = nrf_cloud_send(&output);
+	} else if (IS_ENABLED(CONFIG_NRF_CLOUD_REST)) {
+		/* send to d2c topic, unless bulk is true, in which case d2c/bulk is used */
+		err = nrf_cloud_rest_send_device_message(context.rest_ctx,
+							 context.device_id,
+							 output.data.ptr, false, NULL);
+	} else {
+		err = -ENODEV;
+	}
+	if (!err) {
+		stats.lines_sent += num_msgs;
+		stats.bytes_sent += output.data.len;
+	}
+
+cleanup:
+	if (base64_buf) {
+		k_free(base64_buf);
+	}
+	ret = ring_buf_get_finish(&log_nrf_cloud_rb, stored);
+	ring_buf_reset(&log_nrf_cloud_rb);
+	num_msgs = 0;
+
+	if (ret) {
+		LOG_ERR("Error finishing ring buffer: %d", ret);
+		err = ret;
+	}
+	return err;
+}
+
+static int logger_out(uint8_t *buf, size_t size, void *ctx)
+{
+	ARG_UNUSED(ctx);
+	int err = 0;
+	struct nrf_cloud_data data;
+	size_t orig_size = size;
+	uint32_t stored;
+	int extra;
+
+	if (!size) {
+		return 0;
+	}
+
+	if (log_format_current == LOG_OUTPUT_TEXT) {
+		extra = 2;
+		if (context.src_name) {
+			int len = strlen(context.src_name);
+
+			/* We want to pass the log source as a separate field for cloud filtering.
+			 * Because the Zephyr logging subsystem prefixes the log text with
+			 * 'src_name: ', remove it before encoding the log.
+			 */
+			if ((strncmp(context.src_name, (const char *)buf, MIN(size, len)) == 0) &&
+			    ((len + 2) < size) && (buf[len] == ':') && (buf[len + 1] == ' ')) {
+				buf += len + 2;
+				size -= len + 2;
+			}
+		}
+		err = nrf_cloud_encode_log(&context, buf, size, &data);
+		if (err) {
+			LOG_ERR("Error encoding log: %d", err);
+			goto end;
+		}
+		log_sequence++;
+	} else {
+		extra = sizeof(struct nrf_cloud_bin_hdr);
+		data.ptr = buf;
+		data.len = size;
+	}
+
+	stats.lines_rendered++;
+	stats.bytes_rendered += data.len;
+
+	do {
+		/* If there is enough room for this rendering, store it and leave. */
+		if (ring_buf_space_get(&log_nrf_cloud_rb) > (data.len + extra)) {
+
+			if (num_msgs == 0) {
+				/* Insert start of buffer marker */
+				if (log_format_current == LOG_OUTPUT_TEXT) {
+					/* Open JSON array */
+					ring_buf_put(&log_nrf_cloud_rb, "[", 1);
+				} else {
+					struct nrf_cloud_bin_hdr hdr;
+
+					hdr.magic = NRF_CLOUD_BINARY_MAGIC;
+					hdr.format = NRF_CLOUD_DICT_LOG_FMT;
+					hdr.ts = context.ts;
+					hdr.sequence = context.sequence;
+					log_sequence++;
+					ring_buf_put(&log_nrf_cloud_rb,
+						     (const uint8_t *)&hdr, sizeof(hdr));
+				}
+			}
+			stored = ring_buf_put(&log_nrf_cloud_rb, data.ptr, data.len);
+			if (stored != data.len) {
+				LOG_WRN("Stored:%u, put:%u", stored, data.len);
+			}
+			num_msgs++;
+			if (log_format_current == LOG_OUTPUT_TEXT) {
+				cJSON_free((void *)data.ptr);
+			}
+			/* Stored rendered output in ring buffer, so we can exit now */
+			break;
+		}
+
+		/* Low on space, so send everything. */
+		err = send_ring_buffer();
```

Comment:
Maybe it would be better to drop any lines past this point, rather than block the logging API on sending the ring buffer

That way if I send 1000 messages, we'll drop 900 of them, rather than send 100, drop 1, send another 100, drop another 1, so on

Reply:
The dropping is handled by the logging system before it gets to `logger_process()`.

Reply:
I misread the code. On re-reading, it will re-try sending the same data, rather than drop the data

Actually, I'm realizing that if `data.len + extra` is larger than the entire ring buffer, this code will hang. Ideally that shouldn't happen, but it might be a good idea to add a breakout for that

Reply:
It is not a problem to block in the output function.  The UART backend does this, for example.

The log output buffer size is quite a bit smaller than the ring buffer size, so it's not possible to have `data.len + extra` larger than the ring buffer.

I do need to add an assert to ensure it is configured this way.

Reply:
>  The UART backend does this, for example

I only ask because this is a network operation rather than a hardware operation, so it could take significantly longer, for instance if a connection times out. I don't actually know what thread is managing the logging back-end, though. If it's the same thread responsible for the system workqueue, though, then it's almost certainly a problem

Reply:
The logging system has its own thread.

Zephyr has a network backend already.  It's quite complex but doesn't meet our needs.  Anyway, it does not appear to do anything special to prevent blocking.